[wg/did] Did wg 2023 team proposal

pchampin opened the pull request:
This PR aims to takes account most of the remarks made during the AC review and the TPAC discussions.

[preview](https://htmlpreview.github.io/?https://raw.githubusercontent.com/w3c/charter-drafts/did-wg-2023-team-proposal/2023/did-wg.html) | [diff](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fw3c.github.io%2Fcharter-drafts%2F2023%2Fdid-wg.html&doc2=https%3A%2F%2Fraw.githubusercontent.com%2Fw3c%2Fcharter-drafts%2Fdid-wg-2023-team-proposal%2F2023%2Fdid-wg.html)

pchampin committed:
address #447

pchampin committed:
remove DID methods, and make DID resolution a REC track deliverable

addresses
- #427 (DID resolution on REC track)
- #431 (remove DID methods)
- #434 (now moot, as DID methods have been removed)

pchampin committed:
add specific exit criteria discussed during TPAC

iherman commented on 2023/did-wg.html:
```diff
@@ -151,12 +150,10 @@ <h2>Motivation and Background</h2>
         <p>
-The Working Group will also seek consensus around DID Method specifications, and
-may produce related documents.
+The Working Group will also take over the work started by the Credenials Community Group around <a href="https://w3c-ccg.github.io/did-resolution/">DID Resolution</a>, to define common requirements, algorithms including their inputs and results, architectural options, and various considerations for the DID resolution and DID URL dereferencing processes.
```
```suggestion
The Working Group will also take over the work started by the Credentials Community Group around <a href="https://w3c-ccg.github.io/did-resolution/">DID Resolution</a>, to define common requirements, algorithms including their inputs and results, architectural options, and various considerations for the DID resolution and DID URL dereferencing processes.
```

iherman commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,18 @@ <h2>Success Criteria</h2>
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected to define a dummy DID method meant only for demonstrating interoperability across independent implementations. While required for all complying implementations, this method wil ...
+    It is also expected that the Working Group can exhibit that at least two independant open specification of DID methods relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```
```suggestion
    It is also expected that the Working Group can exhibit that at least two independent open specifications of DID methods relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```

thereisnogabe reviewed (approved):
Block/TBD is supportive of these changes

OR13 commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,18 @@ <h2>Success Criteria</h2>
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected to define a dummy DID method meant only for demonstrating interoperability across independent implementations. While required for all complying implementations, this method wil ...
+    It is also expected that the Working Group can exhibit that at least two independant open specification of DID methods relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```
Isn't this required for all TR track documents? 

Specifically all normative statements need to be tested by 2 independent implementations.

I don't see why this sentence is needed, and it could be misread as "the working group will endorse 2 specific did methods (maintained outside the wg) that are suitable for production"... thats twice the problem of standardizing a specific did method.

iherman replied:
> Isn't this required for all TR track documents?
> 

Not necessarily. The requirements for guidelines (like WCAG) or vocabularies may be different. Hence, the requirement is that a charter must explicitly define its minimal CR requirement.

pchampin replied:
> it could be misread as "the working group will endorse 2 specific did methods (maintained outside the wg) that are suitable for production"

I don't think there is any assumption that a WG endorse the implementations its mentions in its implementation report. Similarly, this risk of misreading seems a bit remote. However, feel free to propose a better wording if you think it can be made clearer.

pchampin replied:
I proposed a new wording of this requirement in 1ecaa35.
@OR13 does it address your concern? If so, are we good to close this conversation?

pchampin replied:
Resolving, as the conversation is being continued [here](https://github.com/w3c/charter-drafts/pull/448#discussion_r1350767221)

OR13 commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,18 @@ <h2>Success Criteria</h2>
 Proposed Recommendation, each normative specification must have an open
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected to define a dummy DID method meant only for demonstrating interoperability across independent implementations. While required for all complying implementations, this method wil ...
```
I don't see why we need to define a dummy did method, just use real URLs for real controller documents, like the W3C VCWG is already proposing to do: https://w3c.github.io/vc-data-integrity/#controller-documents

See the first example in the spec:

https://w3c.github.io/vc-data-model/#example-a-simple-example-of-a-verifiable-credential

```json
{
"id": "https://university.example/issuers/565049",
"verificationMethod": [{
  "id": "https://university.example/issuers/565049#key-123",
  "type": "Multikey",
  "controller": "https://university.example/issuers/565049",
  "publicKeyMultibase": "z6MkmM42vxfqZQsv4ehtTjFFxQ4sQKS2w6WR7emozFAn5cxu"
  }]
}
```

OR13 replied:
I guess this raises the question of if HTTPS controller documents require a "DID Method Spec"... I don't think they do, so the prohibition on defining specific DID Methods does not apply to HTTPS controller documents.

OR13 replied:
See related issue currently blocking the VCWG: https://github.com/w3c/vc-jose-cose/issues/160

pchampin replied:
@OR13 I'm not sure I fully understand your argument here.
> just use real URLs for real controller documents

My understanding is that DID resolvers are expecting DIDs, nor arbitrary URIs, so using "real URLs" would be out of scope. Of course, a straightforward way to define the "dummy did method" would be to naively encode a real URL in the DID.

OR13 replied:
Thats what I am saying. The concept of "controller documents" has already expanded dids and did documents to URLs in the verifiable credentials working group.

If the DID WG does not intend to harmonize the concept of "controller document" resolution with the concept of "did document" resolution, I object to the charter entirely.

Its my understanding that is the current intention (among the authors of controller document section of the data integrity proof spec, and the authors of the did core / did resolution specs).

rxgrant replied:
@OR13 DID Resolution is about more than what a controller document looks like.  That is a data model concern that doesn't solve the problem at hand.

pchampin replied:
@OR13 you wrote
> If the DID WG does not intend to harmonize the concept of "controller document" resolution with the concept of "did document" resolution, I object to the charter entirely.

I don't see any need for harmonization, as I see is a clear separation of concerns between what vc-data-integrity says about controller document resolution and what this charter says about DID resolution. More specifically, [vc-data-integrity, Sec 4.7](https://www.w3.org/TR/vc-data-integrity/#retrieve-verification-method) says (emphasis is mine):

> Let controllerDocument be the result of dereferencing controllerDocumentUrl, **according to the rules of the URL scheme** and using the supplied options. 

My reading is that, if the URL starts with `http:` or `https:` implementers should apply RFC9110; if it starts with `did:`, implementers should apply DID Resolution; etc. @msporny please chime in if my understanding is incorrect.

To summarize:
* vc-data-integrity defers to the specs governing each URL scheme for dereferencing
* DID Resolution aims to govern the dereferencing `did:` URLs

msporny replied:
@pchampin wrote:
> My reading is that, if the URL starts with http: or https: implementers should apply RFC9110; if it starts with did:, implementers should apply DID Resolution; etc. @msporny please chime in if my understanding is incorrect.

Yes, that is the correct interpretation, IMHO, @pchampin.

peacekeeper replied:
> use real URLs for real controller documents

Is the notion of "controller document" compatible with "DID document"? I hope so... If not, I don't see how DID Resolution could fix any potential incompatibilities, since DID Resolution isn't about (re-)defining those data models...

msporny replied:
A DID Document is a specific profile of a Controller Document.

pchampin replied:
Resolving this thread as it is made "outdated" by d8df047. @OR13 feel free to make a new comment if you feel that the "controller document" still needs dicusssion.

OR13 commented on 2023/did-wg.html:
```diff
@@ -360,9 +343,6 @@ <h2>Success Criteria</h2>
     </p>
 	
-    <!-- Relevance and momentum -->
-	  <p>All new features should be supported by at least two intents to implement before being incorporated in the specification.</p>
```
I don't like how this moved here: https://github.com/w3c/charter-drafts/pull/448/files#r1343224378

pchampin replied:
This wording came directly from the charter template, from which it will probably be removed soon (see https://github.com/w3c/charter-drafts/pull/437). The notion of "intent to implement" is quite browser specific, and therefore does not apply here.

The other changes in this commit are meant to cover the same general idea, while being more adapted and more specific to this WG.

OR13 replied:
done: https://github.com/w3c/charter-drafts/pull/448/files#r1350767221

pchampin replied:
> done: https://github.com/w3c/charter-drafts/pull/448/files#r1350767221

Sorry, I don't understand this comment... and the anchor does not seem to link to any specific place in the file...

OR13 replied:
Not sure, seems like a copy paste bug, this can be resolved.

OR13 commented on 2023/did-wg.html:
```diff
@@ -151,12 +150,10 @@ <h2>Motivation and Background</h2>
         <p>
-The Working Group will begin working toward a specification for DID Resolution
-and DID URL Dereferencing.
+It is expected that these two goals will be taken by distinct task forces in the WG.
```
I would prefer to not have 2x meetings... hopefully thats not what is being implied by this sentence.

iherman replied:
The charter does not say that.

pchampin replied:
@OR13 To be honest, I would also prefer to not have 2x meetings :wink: but as @iherman points out, the charter makes no assumption either way. Charters usually don't go into the detail of regular group meetings, this is up to the chairs to decide what's the best meeting frequency for each task-force is.

Are you ok with resolving this conversation?

rxgrant replied:
I believe that, as discussed at TPAC, double the meetings is not implied or in any way acceptable to the DID-WG.

pchampin replied:
seeing no protest, I resolve this conversation

OR13 commented on 2023/did-wg.html:
```diff
@@ -182,10 +179,7 @@ <h2>Scope</h2>
-Working Drafts for DID Resolution or DID URL Dereferencing, or of DID Method
-Specifications. Any new documents intended for Recommendation will be
-limited to the Working Draft maturity level.
+The Working Group may also publish new Working Group Notes.
```
Isn't this always true?

iherman replied:
It is not 100% clear. The only thing the [process document](https://www.w3.org/2023/Process-20230612/#note-track) says is that (paraphrasing the reference above) WG Notes may be published by WGs only (as opposed to other types of groups), but that does not say that a specific group is indeed allowed to do so without further ado. To avoid further unnecessary discussions, I think it is wise to keep this in the charter.

pchampin replied:
Nothing to add to what @iherman wrote.

@OR13 any objection to resolving this conversation?

OR13 replied:
I'd prefer not to give the working group a blank check on notes, based on the previous experience we have had, the notes become a dumping ground for lack of working group consensus. 

That in turn harms the quality of the other deliverables by comparison to the notes. 

For example, a did method could be published as a note, and then later made a recommendation.

I think it's better to have a charter that is narrow, specific, and where each item has the full support of the working group.

Allowing fractions of the group to publish notes that don't reflect working group consensus seems like something that will harm our primary objective, and we have an active CG that can publish community drafts, which don't require full working group consensus.

iherman replied:
> Allowing fractions of the group to publish notes that don't reflect working group consensus seems like something that will harm our primary objective, and we have an active CG that can publish community drafts, which don't require full working group consensus.

I do not understand this argument. The very act of publishing a note ***does*** require group consensus, i.e., a  formal resolution of the Working Group.

rxgrant replied:
I agree with @OR13.  I see no need for more of the disasterous notes process in this WG.

rxgrant replied:
> I do not understand this argument. The very act of publishing a note _**does**_ require group consensus, i.e., a formal resolution of the Working Group.

@iherman, you are hearing that members wish to resolve that matter of consensus now, by stating that, based on prior experience, it will not exist for Notes.

iherman replied:
@rxgrant, I interpret what I see in the comments as a fundamental mistrust in the management of the group with regard to the consensus seeking process, at least for the acceptance of notes being published. I sincerely hope I completely misunderstand this.

However. If the intention is to strongly limit the ability of this group to publish unforeseen notes, then I believe there are two things that must be done:

1. List in the deliverables _what_ notes are allowed to be published
2. Add to the 'out of scope' section that the group is _not_ allowed to publish anything else, including wg notes.

The reason the second point is relevant is that such a restriction is highly unusual in working groups and, therefore, should be made explicit.

But it may well be (and I hope!) that I misunderstand the discussion here.

iherman replied:
One more remark:

@OR13 said:

> For example, a did method could be published as a note, and then later made a recommendation.

The process makes this more difficult, actually. The latest version says:

> A Group Note (NOTE) is published to provide a stable reference for a useful document that is not intended to be a formal standard. [See [2023 process document](https://www.w3.org/2023/Process-20230612/#Note)]

Turning a note into a Recommendation would require a separate chartering step for a new WG and the AC might as well refuse to accept this if it simply comes directly from a Note.

rxgrant replied:
> @rxgrant, I interpret what I see in the comments as a fundamental mistrust in the management of the group with regard to the consensus seeking process, at least for the acceptance of notes being published. I sincerely hope I completely misunderstand this.

@iherman  It is absolutely approriate to define where consensus will be reached in the chartering process.  If you think chartering is not part of the consensus process, and that it only begins later, then I hope that you will clarify further.

iherman replied:
@rxgrant I believe you misunderstood. I was referring to the fact that whether a WG decides to publish (or not) a particular WG Note is subject to group consensus. Therefore, I am not sure that it is worth saying more than the proposed generic text in the charter, namely that the WG may publish WG notes.

If the group has a consensus in making the charter more strict, then be it.

peacekeeper replied:
I don't really have a strong opinion whether notes should be in scope or not for this WG, I'm fine either way. But referring to them as a "dumping ground" or "disastrous process" isn't helpful. Which note(s) from a past WG are you referring to, and what was wrong in those cases?

OR13 replied:
The DID Spec Registries is a WG note.

I don't believe it has gone well.

Almost every new method registered has JSON-LD / RDF issues with it.

But the registration process does not require well formed documents.

Perhaps the new process has addressed the previous challenges the DID WG faced, which led us to publish the JSON-LD context, JSON-LD extension points, and the DID Method Registry as a WG note.

I'm sorry my maintainer experience here requires me to describe the note as a "dumping ground", but thats what it is, since the maintainers of the registry have little ability to ensure quality, and the working group is not "coming to consensus" to merge the PRs to this note... Its rare that we get more than 1 or 2 reviews of a new PR here.

peacekeeper replied:
> The DID Spec Registries is a WG note.

I think the method registry part has indeed been problematic at times and definitely a lot of work for maintainers. But other parts like extension parameters and extension metadata properties have been really useful in my opinion...

rxgrant replied:
@iherman 

> > Allowing fractions of the group to publish notes that don't reflect working group consensus seems like something that will harm our primary objective, and we have an active CG that can publish community drafts, which don't require full working group consensus.
> 
> I do not understand this argument. The very act of publishing a note _**does**_ require group consensus, i.e., a formal resolution of the Working Group.

Do you agree or disagree with [this statement](https://github.com/w3c/did-imp-guide/pull/27/#issuecomment-918745249) by the Chair, which states that the contents of Notes is not subject to group consensus?

iherman replied:
@rxgrant,

I believe you compare two different things. And I realize that this is all a bit legalistic.

- To be precise, what I said is that it requires a Working Group consensus to register a completely new entry in the formal collection of W3C's publications, a.k.a. a document with a `https://www.w3.org/TR/XXX` URL, where `XXX` is what the jargon calls the "short name" of the publication. A good example is the [CBOR Note](https://www.w3.org/TR/did-cbor-representation/) whose short name is `did-cbor-representation`. This required a [formal resolution](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-06-15-did#resolution1) of the WG on its [meeting on the 15th of June '21](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-06-15-did#section2).

    Note that the resolution also included the approval for the usage of "echidna" for automatic publishing, ie, the publication of a new version of that Note as a result of a github PR. That is possible because, though each version of that Note is published as a "new" Note (as @brentzundel said in https://github.com/w3c/charter-drafts/pull/448#discussion_r1378653120), it is not a new entry in the W3C publication list, hence this automatic update is possible as far as W3C is concerned. (The group could have decided to require a formal WG resolution for the publication of each new version. It didn't.)
- _The content_ of the note does not require 100% consensus for all parties involved in the WG; this depends on the nature of the Note itself. The requirements are fairly loose, although it is indeed good practice to seek for consensus. (After all, per [process](https://www.w3.org/2023/Process-20230612/#note-track), "A Group Note (NOTE) is published to provide a stable reference for a useful document ***that is not intended to be a formal standard.***" Emphasis is mine.)

rxgrant replied:
@peacekeeper

> I don't really have a strong opinion whether notes should be in scope or not for this WG, I'm fine either way.  But referring to them as a "dumping group" or "disastrous process" isn't helpful.  Which note(s) from a past WG are you referring to, and what was wrong in those cases?

The implementation guide literally says that secp256k1, clearly the most battle-hardened cryptographic curve in the world, is unfit for use.  I gave it [an effort](https://github.com/w3c/did-imp-guide/pull/36) on clearly superior technical grounds and the editor tried to say that there was no grounds to delete any submission in the Note.  As I noted in the discussion at the time, that is an instance of [teach the controversy](https://en.wikipedia.org/wiki/Teach_the_Controversy).

While I **am disappointed** in the editor's choice to not seek consensus with the weakest technical objections, I do **not even blame** them for their [quality-destroying definition of consensus](https://github.com/w3c/did-imp-guide/pull/27/#issuecomment-916281723).  Others in the group promoted Notes as a dumping ground for all topics that could not reach consensus.  There was an absolutely toxic playgroud created for bullying and making unsubstantiated disparaging remarks, and it is no surprise to see it taken advantage of without regard to any consensus-seeking principle.

**edits:** This comment included an overstatement about whether the policy was applied evenly.  Upon reviewing further information, I believe the policy was applied evenly.

rxgrant replied:
The above https://github.com/w3c/charter-drafts/pull/448#discussion_r1381888017 is edited to remove an overstatement about whether the policy was applied evenly. 

**edit:** and again edited to not unduly blame the editor.

pchampin replied:
Resolving this conversation, as the discussion about notes has drifted into other conversations in this PR.

pchampin committed:
Update 2023/did-wg.html

Co-authored-by: Ivan Herman <ivan@w3.org>

pchampin committed:
Update 2023/did-wg.html

Co-authored-by: Ivan Herman <ivan@w3.org>

OR13 commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,18 @@ <h2>Success Criteria</h2>
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected to define a dummy DID method meant only for demonstrating interoperability across independent implementations. While required for all complying implementations, this method wil ...
+    It is also expected that the Working Group can exhibit that at least two independent open specifications of DID methods relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```
```suggestion
    It is also expected that the Working Group can exhibit that at least two independent open specifications for each feature defined by <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```

pchampin replied:
the "two independant implementation for each feature" is already mentioned in the previous paragraph of the charter (applying to "each normative specification", so that includes DID Resolution).

the paragraph you are proposing to change is about something else.

pchampin replied:
I proposed a new wording of this requirement in 1ecaa35.
@OR13 does it address your concern? If so, are we good to close this conversation?

OR13 replied:
I suggest removing this paragraph then, since the features will be tested.

If we are not standardizing DID Methods, we don't need "2 independent did methods".

pchampin replied:
oops... 1ecaa35 had a typo ("begin" instead of "being") that I just fixed;



> I suggest removing this paragraph then, since the features will be tested.
> If we are not standardizing DID Methods, we don't need "2 independent did methods".

This requirement was the result of the discussions at TPAC. Requiring the *existence* of DID methods developed *outside* of the group is the middle ground where the "pro-methods" and the "anti-methods" agreed to meet.
Would this be a blocker for you?

OR13 replied:
I'd argue its already satisfied, I approve new did methods all the time over here, see the history:
https://github.com/w3c/did-spec-registries

We also have more than 2 did methods under test... in the current test suite, which tests the result of resolution and dereferencing....

It's a blocker for me... I don't think its doing anything useful, and I can imagine it doing a lot of harm.

As an alternative, we could try to standardize a specific did method by itself, and drop the community registry, but that seems equally unlikely to pass.

The problem is creating the appearance of W3C endorsement for a subset of methods, while trying to create something decentralized... endorsement of specific methods by W3C defeats the purpose of the work.

If we are trying to demonstrate that methods exist... it seems like we don't need to put that in the charter at all.

rxgrant replied:
I agree with @OR13 that the paragraph should be mostly removed.  See proposal [here](https://github.com/w3c/charter-drafts/pull/448/files#r1370382469).

pchampin replied:
> I agree with @OR13 that the paragraph should be mostly removed. See proposal [here](https://github.com/w3c/charter-drafts/pull/448/files#r1370382469).

@rxgrant The paragraph you are suggesting to remove in the proposal to your point to is not the one under discussing here.

rxgrant replied:
@pchampin  I would agree with you that line 319 is the one I'm proposing the most significant revision to, and that your edits to line 320 (incorporated into my proposal and not removed as might be inferred from my comment above) address a reasonable requirement to provide evidence of interoperability work being done elsewhere.  

I would also agree with you that there was redundancy in @OR13's initial suggestion, since the paragraph above does cover features.  

I would disagree with you about whether that makes the paragraph you pointed at the one we are discussing.  All edits (yours, mine, and Orie's) are proposed on lines 319 and 320, which are one paragraph.

I am also agreeing with @OR13's [suggestion to remove the paragraph under discussion](https://github.com/w3c/charter-drafts/pull/448#discussion_r1363840729) that DID-WG should not be in the business of standardizing DID Methods.

pchampin replied:
Resolving this thread as it is made "outdated" by d8df047

pchampin committed:
rephrase the requirement for two independant DID methods

pchampin committed:
typo

rxgrant commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,18 @@ <h2>Success Criteria</h2>
 Proposed Recommendation, each normative specification must have an open
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected to define a dummy DID method meant only for demonstrating interoperability across independent implementations. While required for all complying implementations, this method wil ...
+    It is also expected that the Working Group can provide evidence of at least two independent open specifications of DID methods being developed outside of the Working Group, and relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```
```suggestion
    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected that the Working Group can provide evidence of at least two independent open specifications of DID methods being developed outside of the Working Group, and relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```

rxgrant replied:
While the focus on DID Resolution is a welcome positive change, this proposal is still unacceptable.

How can forcing an insecure dummy DID Method be required for complying implementations yet not used in production?  Are you going to demand that implementors make insecure design decisions if did:dummy demands them?  What are you going to do, throw away the code in production?  What's the test for interoperability in production?

If there is a reasonable technical hurdle here for the DID-WG to clarify conformant implementations, it would be in offering test cases for implementors.  What does a test case look like, if resolvers do not all need to resolve the same DIDs?  Ahh, that might require the working group to meet and think about it.  Hmm, that would require not being in AC-review right now.  Oh, isn't that the substance of one of the formal objections?

One possible answer to what test cases should look like is a _generic resolver API_ that various DID Methods can show conformance to in their own way.  Conformant resolver implementations would publicly share their responses per DID Method, and higher level applications could declare whether they will accept DID Documents produced by those resolvers (whether resolving as remote servers or as compiled-in libraries).

In summary, the DID-WG choosing a dummy DID Method for standardization by this group has all the problems motivating [DCD's original objections](https://lists.w3.org/Archives/Public/public-new-work/2023Sep/0011.html).

The above edits allow the DID-WG to point outside the group to find DID Methods for interoperability.  That is a reasonable way to proceed given the group's truncated powers of creative consensus-building.

pchampin replied:
Resolving this thread, as it is made "outdated" by d8df047

pchampin commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,18 @@ <h2>Success Criteria</h2>
 Proposed Recommendation, each normative specification must have an open
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected to define a dummy DID method meant only for demonstrating interoperability across independent implementations. While required for all complying implementations, this method wil ...
+    It is also expected that the Working Group can provide evidence of at least two independent open specifications of DID methods being developed outside of the Working Group, and relying on <a href="#deliverable-did-core">DID Core</a> and <a href="#deliverable-did-resolution">DID Resolution</a>.
```
```suggestion
    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected that each of the independant implementations mentioned above support at least two DID methods with an open specification (i.e. a specification that is accessible to all for implementation and deployment). It is also expected that each pair of the independant implementations mentioned above support at least one common DID method with an open specification.</p>
```
@OR13  @rxgrant would this change work for you?
It gets rid of the dummy DID method (which, in my understanding) was not a strong requirement from anybody at TPAC, just a suggestion),
and it changes the focus on DID method to consider them as part of the implementations.
It shifts the "choice" of the DID methods further away from the WG towards the implementers, hopefully removing any risk of a perceived endorsement by the WG.

I'm not entirely sure that this change would fly with everyone else involved in the discussion, so I will commit it only if you tell me that it is acceptable for you.

rxgrant replied:
@pchampin This change is acceptable.

rxgrant replied:
@pchampin Why would you default to a version that is definitely unacceptable to me, if others object?  What is your technical reason for doing so?

What might be the technical reason for objecting coming from others?  Are they part of this conversation?

peacekeeper replied:
I also like this change. I think the idea of defining a "dummy DID method" has good intentions, but I'm worried it will distract the work of the group. At TPAC, a strong desire for "practical interoperability" was expressed, and I'm convinced we can easily demonstrate such interoperability without actually defining a DID method in the WG.

msporny replied:
I agree with @peacekeeper's comments and sentiments.

pchampin replied:
> @pchampin Why would you default to a version that is definitely unacceptable to me, if others object? What is your technical reason for doing so?

@rxgrant I never said that I would default to the current version if other people objected to the change above. All I wrote was that I would not make the change if it didn't satisfy you and @OR13  more than the current version. I see no reason (technical or not) to make a change that risks creating more objections if it does not help solving the existing objections...

But in fact, as you confirm that it is acceptable to you, I'll try my luck with this change.

rxgrant commented on 2023/did-wg.html:
```diff
@@ -182,10 +179,7 @@ <h2>Scope</h2>
-Working Drafts for DID Resolution or DID URL Dereferencing, or of DID Method
-Specifications. Any new documents intended for Recommendation will be
-limited to the Working Draft maturity level.
+The Working Group may also publish new Working Group Notes.
```
```suggestion
```

Remove the Notes.  The working group will not publish any new notes.  See also change in "Out of Scope" section.

rxgrant replied:
I cannot suggest changes in the "Out of Scope" section, because this is not my pull request.  @pchampin , please implement @iherman 's suggestion to state that Notes are out of scope.

brentzundel replied:
Making all possible Notes out of scope for the next DID WG would be a severe limitation. I'm not understanding the reasoning behind this?
The nature of a Working Group is to produce Recommendations. Sometimes these Recommendations are benefitted by the additional publication of useful documents that are associated with the Recommendation.
A ban on all Notes would prohibit the next DID WG from publishing anything other than Recommendation track documents. It would formally put out of scope any updates to the DID Method Rubric, DID Use Cases, DID Implementation Guide, etc. 
Without a strong argument from a large group in support of removing Notes from the charter, I could not support this step.

msporny replied:
I agree with @brentzundel. I can't think of a recent W3C WG that was limited in this way from publishing Notes. Digital Bazaar would not support such a limitation because part of what a maintenance WG needs to do is maintain the Notes it has already published... the Use Cases, Rubric, and Implementation Guide being at least among the set of Notes that need to be maintained.

rxgrant replied:
@brentzundel 

>  It would formally put out of scope any updates to the DID Method Rubric, DID Use Cases, DID Implementation Guide, etc.

This is a straw man argument.  I have not proposed a ban on all notes.  I have proposed a ban on new notes.  If there are new notes that you want, let us know now.

@msporny 

You are supporting a straw man argument.  If there are new notes that you want, let us know now.

brentzundel replied:
Unfortunately, W3C Process would consider any update to a previously published Note as a new Note. That's how it works. A ban on new Notes is a ban on all Notes.

This is not a straw man. We don't have some dark hidden agenda. We are simply pointing out some consequences of the changes you are insisting on. 

If new Notes are out of scope for the working group, then it will publish no Notes.

rxgrant replied:
@brentzundel 
> If new Notes are out of scope for the working group, then it will publish no Notes.

In the context of this discussion, @iherman has [already advised](https://github.com/w3c/charter-drafts/pull/448#discussion_r1371934455) listing the Notes to be updated explicitly.  It's completely valid to say that unforseen Notes will not be published.

brentzundel replied:
While it may be valid to say that no unforeseen Notes will be published, I am failing to understand why such a thing would be desirable. 
Manu and I have expressed concerns that a prohibition against all unforeseen Notes (even if an exception is carved out for updates to existing Notes) could potentially cause significant difficulties for the working group. I am not understanding what the benefit of such a prohibition would be, either for the working group or for the progression of DIDs in general. 

@rxgrant Could you please outline your reasons for wanting to prohibit the working group from publishing Notes?

rxgrant replied:
DCD does not approve this charter in [its current form](https://github.com/w3c/charter-drafts/pull/448/commits/d8df0476a32f2ea4331c4f589f4593cf6bcba468).

I would not be opposed to listing a DID Resolution Implementation Guide as a foreseen Note, if the consensus process for editing that Note is written down beforehand and is a major advance from the process used for the DID Implementation Guide, which I have described problems with [here](https://github.com/w3c/charter-drafts/pull/448#discussion_r1381888017).

I would also be in favor of wording that **allows** unforseen Notes _related to DID Resolution_ when that wording allows **no wiggle room** for Notes about DID Methods.

The following must be explicitly prohibited in the Out of Scope section:
- work on particular DID Methods
- publishing DID Methods as Notes

Until I see better wording, I think the best way to achieve this is to hold the line on no unforseen Notes.

brentzundel replied:
I have no problem with the prohibition against DID Methods being extended to Notes. Such a thing seems unnecessary to me, considering that publishing something as a Note that would ever be considered for the Recommendation track is already out of bounds. 
Extending that prohibition against Notes to include all possible unforseen Notes is not something I could support.

peacekeeper replied:
Hmm since DID methods were removed in https://github.com/w3c/charter-drafts/pull/448/commits/aacf57d4815121e3946c0edef0c3ad8d894d0547, would it "hurt" to explicitly mention DID methods as out-of-scope? But not prohibit other Notes? It seems this would address the concerns?

jandrieu replied:
> Such a thing seems unnecessary to me, considering that publishing something as a Note that would ever be considered for the Recommendation track is already out of bounds.

I generally agree with giving the ability for the WG to publish additional notes is useful.

However, I find the characterization that consensus doesn't apply to Notes to be out of order.

Consensus applies to everything the WG does.

Since that seems to be in doubt, I recommend we add that notion explicitly.

pchampin replied:
@rxgrant, my understanding is that [@jandrieu's proposed change](https://github.com/w3c/charter-drafts/pull/448#discussion_r1385594681) adequately addresses your issue about notes (that's how I read [your response](https://github.com/w3c/charter-drafts/pull/448#discussion_r1390086133)). Therefore, I'm resolving this conversation.

pchampin replied:
unresolving this conversation, as [I had misunderstood](https://github.com/w3c/charter-drafts/pull/448#issuecomment-1816289464) @rxgrant's comment. My apologies.

peacekeeper replied:
```suggestion
The Working Group may also publish new Working Group Notes. Working Group Notes will not include DID method specifications. The Working Group however welcomes and may support efforts to standardize DID methods in other Working Groups and/or organizations.
```

@rxgrant and others, how about something like this as an idea?

rxgrant replied:
> The Working Group however welcomes and may support efforts to standardize DID methods in other Working Groups and/or organizations.

It's not okay to me to ask the WG to "support efforts" elsewhere.  That has all the same problems as doing something in the WG!

rxgrant replied:
proposal: Notes should not include submissions for DID Methods and should not address issues regarding any particular DID Method.  The WG welcomes DID Method standardization outside the WG (whether in the W3C or in other standards bodies).

pchampin committed:
Change DID Resolution success criteria

removed the "dummy DID method" and the "provide evidence of existing DID methods"

instead, the "evidence of existing DID methods" is deferred to DID Resolver implementations. Interoperability will be demonstrated by ensuring that resolvers support DID methods in common.

rxgrant reviewed (approved):
I approve the changes regarding using external DID Methods.

**edit:**  As of November 3rd, DCD does **not** approve of the charter in its current form, and [offers suggestions](https://github.com/w3c/charter-drafts/pull/448#discussion_r1381903190) that make it better.

jandrieu commented on 2023/did-wg.html:
```diff
@@ -240,34 +234,18 @@ <h3>
              </p>
              <p class="milestone"><b>Expected completion:</b>
-             It is not intended that DID Resolution and DID URL Dereferencing will advance beyond <a href="https://www.w3.org/Consortium/Process/#RecsWD">Working Draft status</a>.
+             <i class="todo">START+24M</i>.
```
```suggestion
             <i class="todo">START+36M</i>.
```
I don't believe we can get through this work in two years.

iherman replied:
Just an admin remark: if this change is approved and merged, the "metadata" at the beginning of the charter proposal should also be changed (it says, for the end date, "Start date + 2 years").

However, note that, these days, the management is reluctant to allow a charter for three years. The WG might be prepared to say, instead, that it asks for two years with a possibility for an extension at the end of that period; essentially, giving the AC the possibility for a check on whether the WG is really progressing. Something to be discussed in advance...

pchampin replied:
Thanks @iherman, I have nothing to add. @jandrieu, @peacekeeper could you live with that (i.e. keeping the 24M deadline in the proposed charter)?

peacekeeper replied:
Two years with possibility for an extension sounds good to me.

rxgrant replied:
Two years with possibility for an extension sounds okay to me.

jandrieu commented on 2023/did-wg.html:
```diff
@@ -151,12 +150,10 @@ <h2>Motivation and Background</h2>
         <p>
-The Working Group will begin working toward a specification for DID Resolution
-and DID URL Dereferencing.
+It is expected that these two goals will be taken by distinct task forces in the WG.
```
```suggestion
It is expected that these two goals will be taken by distinct task forces in the WG.

The Working Group MUST pursue consensus in any task pursued under the name of the working group--whether by Chairs, Editors, or other contributors--including any Notes for publication by the WG and any Charters for a future DID WG.
```

rxgrant replied:
Along with my proposed changes, I support this change and have budgeted my time and energy to continue to escalate objections against any further work planned by this WG that could _possibly_ abuse process in the ways that I have so painfully observed when its members are not properly constrained.

peacekeeper replied:
I have always assumed that this proposed language applies anyway, but I guess it helps to emphasize it, so I'd support this too.

rxgrant replied:
> I have always assumed that this proposed language applies anyway

I would be willing to resubmit a pull request substantially similar to [DID Implementation Guide pull #36](https://github.com/w3c/did-imp-guide/pull/36) if this WG, under the current heightened scrutiny, agrees that pursuing consensus is required for Notes.

rxgrant replied:
Note that although the email summaries reference the original change proposed by @pchampin , which looks like: 

> It is expected that these two goals will be taken by distinct task forces in the WG.

the current text under discussion is: 

> The Working Group MUST pursue consensus in any task pursued under the name of the working group--whether by Chairs, Editors, or other contributors--including any Notes for publication by the WG and any Charters for a future DID WG.

So if you are only reading along in email then you will be misled.

pchampin replied:
@jandrieu just to be clear about the changes you suggest above :
by "pursue consensus", do you mean to rule out completely recourse to [Section 5.2.2. Managing Dissent](https://www.w3.org/2023/Process-20231103/#managing-dissent) and [Section 5.2.3. Deciding by Vote](https://www.w3.org/2023/Process-20231103/#Votes) of the process?

jandrieu replied:
I do not. I mean to make sure that staff & chairs don't assert that actions taken in the name of the group do not require the seeking of consensus.

jandrieu reviewed (changes requested):
This looks better, but still need help, especially regarding expectations about consensus.

rxgrant commented:
On Fri, Nov 17, 2023 at 8:56 AM Pierre-Antoine Champin
***@***.***> wrote:
> +The Working Group may also publish new Working Group Notes.
>
> @rxgrant, my understanding is that @jandrieu's proposed change adequately addresses your issue about notes (that's how I read your response). Therefore, I'm resolving this conversation.

Pierre-Antoine, this is too aggressive of a resolution.  I referred to
"along with my other changes" and continue to object to work that
leaves room for Notes that may later attempt to pick winners and
losers by publishing documents that involve DID Methods, no matter how
perfect of a consensus definition the WG arrives at for work on future
Notes.

You should instead note the support for limiting the scope of work on
Notes, and work on a draft that achieves that.

pchampin committed:
Brent Zundel is now an IE

jyasskin commented on 2023/did-wg.html:
```diff
@@ -337,13 +315,17 @@ <h2>Success Criteria</h2>
 Proposed Recommendation, each normative specification must have an open
 test suite of every feature defined in the specification.</p>
+
+    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected that each of the independant implementations mentioned above support at least two DID methods with an open specification (i.e. a specification that is accessible to all for imp ...
```
This should set a more specific bar than just "open specification". @cwilso and I suggested over [email](https://lists.w3.org/Archives/Public/public-did-wg/2023Nov/0017.html) that it be something like

```suggestion
    <p>In order for <a href="#deliverable-did-resolution">DID Resolution</a> to advance to <a href="https://www.w3.org/Consortium/Process/#RecsPR" title="Proposed Recommendation">Proposed Recommendation</a>, it is expected that each of the independant implementations mentioned above support at least two DID methods with open specifications at at least the Candidate Recommendation Snapshot level, or a level with equivalent vetting and patent grants at another standards body. It is also expected that each pair of the independant implementations mentioned above support at least one common DID method with an open specification.</p>
```

jandrieu replied:
We would oppose this. 

None of the DID methods are at that level of maturity. 

However, I agree that some further definition of "open" might be useful.

pchampin commented:
Following @iherman 's suggestion at the end of [today's meeting](https://www.w3.org/2023/11/30-did-minutes.html), I merge this PR as is and will make new PRs with the remaining pending modifications.
